debug/thread.go

```diff
@@ -0,0 +1,513 @@
+			return err
+		}
+
+		if a == breakAction {
+			t.logger.Debug("break requested; not unblocking breakpoint latch")
+		} else {
+			t.logger.Debug("No break requested; unblocking breakpoint latch")
+			t.breakpointLatch.unblock()
+		}
+	}
+}
+
+func (t *thread) pause() error {
+	t.logger.Debug("Pausing thread: %d", t.id)
+	t.breakpointLatch.block()
+	return nil
+}
+
+func (t *thread) resume() error {
+	t.logger.Debug("Resuming thread: %d", t.id)
+	t.breakpointLatch.unblock()
+	return nil
+}
+
+func (t *thread) current() (int, *topdown.Event, error) {
+	i, e := t.stack.Current()
+	return i, e, nil
+}
+
+func (t *thread) stepIn() (eventAction, error) {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	if t.stopped {
+		return nopAction, fmt.Errorf("thread stopped")
+	}
+
+	var a eventAction
+	for {
+		i, e := t.stack.Next()
+		t.logger.Debug("Step-in on event: #%d", i)
+
+		var s threadState
+		var err error
+		a, s, err = t.eventHandler(t, i, e, t.state)
+		if err != nil {
+			return nopAction, err
+		}
+		t.state = s
+
+		if a != skipAction {
+			break
+		}
+	}
+
+	return a, nil
+}
+
+func (t *thread) stepOver() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	if t.stopped {
+		return fmt.Errorf("thread stopped")
+	}
+
+	_, startE, err := t.current()
+	if err != nil {
+		return err
+	}
+
+	baseQueryVisited := false
+Loop:
+	for {
+		i, e := t.stack.Next()
+		t.logger.Debug("Step-over on event: #%d", i)
+
+		if e != nil && e.QueryID == 0 {
+			baseQueryVisited = true
+		}
+
+		a, s, err := t.eventHandler(t, i, e, t.state)
+		if err != nil {
+			return err
+		}
+		t.state = s
+
+		if a == skipAction {
+			continue
+		}
+
+		var qid uint64
+		if e != nil {
+			qid = e.QueryID
+		}
+
+		switch {
+		case startE == nil:
+			t.logger.Debug("Resuming on query: %d; first event", qid)
+			break Loop
+		case a == breakAction:
+			t.logger.Debug("Resuming on query: %d; break-action", qid)
+			break Loop
+		case e == nil:
+			t.logger.Debug("Resuming on query: %d; no event", qid)
+			break Loop
+		case e.QueryID == 0:
+			t.logger.Debug("Continuing past query: %d; base-query", qid)
+		case e.QueryID <= startE.QueryID:
+			t.logger.Debug("Resuming on query: %d; start-query: %d", qid, startE.QueryID)
+			break Loop
+		case baseQueryVisited:
+			t.logger.Debug("Resuming on query: %d; base-query visited", qid)
+			break Loop
+		default:
+			t.logger.Debug("Continuing past query: %d", qid)
+		}
+	}
+
+	return nil
+}
+
+func (t *thread) stepOut() error {
+	t.mtx.Lock()
+	defer t.mtx.Unlock()
+
+	if t.stopped {
+		return fmt.Errorf("thread stopped")
+	}
+
+	_, c, err := t.current()
+	if err != nil {
+		return err
+	}
+
+	for {
+		i, e := t.stack.Next()
+		t.logger.Debug("Step-out on event: #%d", i)
+
+		a, s, err := t.eventHandler(t, i, e, t.state)
+		if err != nil {
+			return err
+		}
+		t.state = s
+
+		if a == skipAction {
+			continue
+		}
+
+		var qid uint64
+		if e != nil {
+			qid = e.QueryID
+		}
+
+		// FIXME: Compare against c.ParentID instead?
```

Comment:
Aren't we using similar logic in `stepOver`? What's the concern here?

Reply:
Yes, this isn't the only place where this could be something to consider. Likely, we wouldn't solve anything by doing this, though, as I think we could be "popped out" more than one query level at a time; making the parent ID irrelevant 🤔 .
I'll remove this comment.